### PR TITLE
fix(angular): add support for Angular 14

### PR DESCRIPTION
**Pull request by sean-perkins**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
  - Some docs updates need to be made in the `ionic-docs` repo, in a separate PR. See the [contributing guide](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#modifying-documentation) for details.
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

Applications that are upgrading from Angular 13 to Angular 14 will receive an exception that breaks the entire app experience. 

```
TypeError: resolver.resolveComponentFactory is not a function
```

This is due to the deprecation and removal of `ComponentFactoryResolver` in Angular 14.

Issue URL: #25353

## What is the new behavior?

- Adds backwards compatibility for Angular 13 applications
- Adds compatibility for Angular 14 applications

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

Dev-build: `6.1.9-dev.11654275237.1b595be3`

Angular 14 demo: https://github.com/sean-perkins/angular-14-demo

For Angular 13, you can use the starters and install the dev-build.

Note: Standalone components are not supported in this phase of development. As far as I can tell, they require `EnvironmentInjector` and other essential infrastructure only available in the Angular 14 bundle. More exploration is required to pass the application's injector through to Ionic's module.

### CODE REVIEW

File: angular/src/directives/navigation/ion-router-outlet.ts [hunk 1]
```diff
@@ -206,7 +207,7 @@ export class IonRouterOutlet implements OnDestroy, OnInit {
     }
   }
 
-  activateWith(activatedRoute: ActivatedRoute, resolver: ComponentFactoryResolver | null): void {
+  activateWith(activatedRoute: ActivatedRoute, resolverOrInjector?: ComponentFactoryResolver | null): void {
```
Comment (amandaejohnston) [inline angular/src/directives/navigation/ion-router-outlet.ts:210]: Should this be `ComponentFactoryResolver | EnvironmentInjector | null`?

File: angular/src/directives/navigation/ion-router-outlet.ts [hunk 2]
```diff
@@ -240,8 +238,33 @@ export class IonRouterOutlet implements OnDestroy, OnInit {
...
 
+      if (resolverOrInjector && isComponentFactoryResolver(resolverOrInjector)) {
+        // Backwards compatibility for Angular 13 and lower
+        const factory = resolverOrInjector.resolveComponentFactory(component);
+        cmpRef = this.activated = this.location.createComponent(factory, this.location.length, injector);
+      } else {
+        /**
+         * Angular 14 and higher.
+         *
+         * When we drop < Angular 14, we can replace the following code with:
```
Comment (amandaejohnston) [inline angular/src/directives/navigation/ion-router-outlet.ts:250]: Could we add a tech debt ticket here too? (Could be the same ticket as the other one)

Also, if/when we don't support Angular 13, do we still need the backwards compatibility code above?

File: angular/src/providers/angular-delegate.ts [hunk 3]
```diff
@@ -96,14 +98,29 @@ export const attachView = (
...
-  const componentRef = location
-    ? location.createComponent(factory, location.length, childInjector)
-    : factory.create(childInjector);
+
+  if (resolverOrInjector && isComponentFactoryResolver(resolverOrInjector)) {
+    // Angular 13 and lower
+    const factory = resolverOrInjector.resolveComponentFactory(component);
+    componentRef = location
+      ? location.createComponent(factory, location.length, childInjector)
+      : factory.create(childInjector);
```
Comment (amandaejohnston) [inline angular/src/providers/angular-delegate.ts:112]: Will we still need this when we drop Angular 13 support? If not, could add the tech debt ticket comment here too.

File: angular/src/providers/angular-delegate.ts [hunk 4]
```diff
@@ -165,3 +182,7 @@ const getProviders = (params: { [key: string]: any }) => {
 const provideNavParamsInjectable = (params: { [key: string]: any }) => {
   return new NavParams(params);
 };
+
+function isComponentFactoryResolver(item: any): item is ComponentFactoryResolver {
+  return !!item.resolveComponentFactory;
+}
```
Comment (amandaejohnston) [inline angular/src/providers/angular-delegate.ts:188]: This function is duplicated in `ion-router-outlet.ts`; would it be possible to make a shared util instead?

**Review by amandaejohnston — COMMENTED** [review]
Great work on this! Just had a few minor things.

File: angular/src/providers/angular-delegate.ts (hunk 3 above)
Comment (sean-perkins) [reply angular/src/providers/angular-delegate.ts:112]: TBD. The standalone components are in dev preview, so Angular is still reserving the right to completely remove them if they discover a reason why they shouldn't exist. As far as I can tell, all of this is backwards support for the previous model of using the component factory resolver for creating the component to resolve for a route. It is still present in Angular's implementation of `router-outlet`, so we would likely want to maintain it until they remove it, which would be Angular 15 at the earliest.

File: angular/src/directives/navigation/ion-router-outlet.ts [hunk 5]
```diff
@@ -229,9 +234,6 @@ export class IonRouterOutlet implements OnDestroy, OnInit {
       const snapshot = (activatedRoute as any)._futureSnapshot;
       // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
       const component = snapshot.routeConfig!.component as any;
-      resolver = resolver || this.resolver;
```
[the remark this answers is not in the document]
Comment (sean-perkins) [reply angular/src/directives/navigation/ion-router-outlet.ts:232]: It looked to be an unnecessary fallback added when stack navigation was added to Ionic Angular: https://github.com/ionic-team/ionic-framework/commit/726938f67cf73d5dcc7cd5c0d2dbbf3b02d91009#diff-676a7dea0a23c6e3559412b3ce504dd48f32062d1756f74aac7f7d72e8d411ac

In Angular's router implementation, the resolver (or injector starting in v14), is always passed through the function: https://github.com/angular/angular/blob/main/packages/router/src/directives/router_outlet.ts#L265-L291. It never uses the instance from where the router outlet is constructed in the DI tree.

**Review by liamdebeasi — APPROVED** [review]
Great job!

**Review by amandaejohnston — APPROVED** [review]

**Pull request merged**